lib/vrl/stdlib/src/for_each.rs

```diff
@@ -0,0 +1,189 @@
+                result: Ok("9"),
+            },
+            Example {
+                title: "recursively iterate array",
+                source: r#"count = 0; for_each([1,2,[3,4]], recursive: true) -> |index, value| { count = count + index + (int(value) ?? 0) }; count"#,
+                result: Ok("14"),
+            },
+        ]
+    }
+
+    fn compile(
+        &self,
+        _state: (&mut state::LocalEnv, &mut state::ExternalEnv),
+        _ctx: &mut FunctionCompileContext,
+        mut arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let recursive = arguments.optional("recursive");
+        let closure = arguments.required_closure()?;
+
+        Ok(Box::new(ForEachFn {
+            value,
+            closure,
+            recursive,
+        }))
+    }
+
+    fn closure(&self) -> Option<closure::Definition> {
+        use closure::{Definition, Input, Output, Variable, VariableKind};
+
+        Some(Definition {
+            inputs: vec![Input {
+                parameter_keyword: "value",
+                kind: Kind::object(Collection::any()).or_array(Collection::any()),
+                variables: vec![
+                    Variable {
+                        kind: VariableKind::TargetInnerKey,
+                    },
+                    Variable {
+                        kind: VariableKind::TargetInnerValue,
+                    },
+                ],
+                output: Output::Kind(Kind::any()),
+                example: Example {
+                    title: "iterate array",
+                    source: r#"for_each([1, 2]) -> |index, value| { .foo = to_int!(.foo) + index + value }"#,
+                    result: Ok("null"),
+                },
+            }],
+            is_iterator: true,
+        })
+    }
+
+    fn call_by_vm(&self, _ctx: &mut Context, _args: &mut VmArgumentList) -> Result<Value> {
+        // TODO: this work will happen in a follow-up PR
+        Ok(Value::Null)
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ForEachFn {
+    value: Box<dyn Expression>,
+    recursive: Option<Box<dyn Expression>>,
+    closure: FunctionClosure,
+}
+
+impl Expression for ForEachFn {
+    fn resolve(&self, ctx: &mut Context) -> Result<Value> {
+        let recursive = match &self.recursive {
+            None => false,
+            Some(expr) => expr.resolve(ctx)?.try_boolean()?,
+        };
+
+        let value = self.value.resolve(ctx)?;
+        let top_level_kind = value.kind();
+
+        let mut iter = value.into_iter(recursive);
+
+        for item in iter.by_ref() {
+            match item {
+                // NOTE:
+                //
+                // `for_each` currently only supports recursively reading nested
+                // key/value pairs if the top-level type is an object. It does
+                // still recurse "through" arrays though, but the array values
+                // aren't returned during iteration.
```

Comment:
I've been going back-and-forth on this. I'm actually leaning towards the other direction after posting this PR, as it seems to be the least surprising behaviour, even if it means you get less specific type information.

Reply:
Hmm, very interesting question. My first gut reaction would agree with not iterating through arrays when top-level is an object.

Since you didn't mention that explicitly: Should the same idea apply for arrays? Should we iterate through nested objects when iterating an array? I'd also lean to "no" in that case.

I'd also add tests for those combinations, to show that we're intentional about that behavior

Reply:
To be clear, what currently happens is that we do iterate _through_ all collection types, but only return key/value if the top-level is an object, and index/value if the top-level is an array.

What I'm proposing, is that we might want to always return all elements in the collection, so both key/value and index/value, even if that means the type definition of the "key or index" variable will almost always be a combination of string and integer, and thus potentially more type coercion/checking is required by the operator within the closure block.

Reply:
As discussed during check-in, given that we don't actually have any real-world use-cases that we collected from the community, needing recursion from this function, we're going to remove that option for the initial release, and defer any question about mixed/non-mixed recursion to when we do get requests for recursion support for `for_each`.

Reply:
resolved in [3dec93f](https://github.com/vectordotdev/vector/pull/12382/commits/3dec93f7c11ac3054a1f722ce0cb601335fbfe7c).